### ISSUE TITLE
Compiler server takes 30 seconds per project

### ISSUE THREAD

**Issue by amcasey** [dotnet/roslyn#6827]
1. Create a new Console Application
 1. From the command line, ```msbuild ConsoleApplication1.sln /t:Rebuild```

Expected: less than 1 second
Actual: 30 seconds

If you disable the compiler server (can't recall the msbuild property name), it takes less than a second.

If you add a second project to the solution, it takes 60 seconds - 30 seconds per project.

**Comment by amcasey** [comment]
FYI @agocke @jaredpar @KevinH-MS

**Comment by amcasey** [comment]
We've been seeing a lot of dependency resolution issues recently, but this doesn't seem to be one.  If I attach to the process, the only exceptions are some initial `IOException`s about the pipe not being ready.

I have a perf trace, but I've been unable to determine what the process is blocking on.

**Comment by jaredpar** [comment]
I'm not seeing this locally.  I followed your repro steps and got the following:

```
C:\users\jaredpar\documents\visual studio 2015\Projects\ConsoleApplication23\ConsoleApplication23> measure-command { msbuild /t:rebuild }


Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 0
Milliseconds      : 418
Ticks             : 4182458
TotalDays         : 4.84080787037037E-06
TotalHours        : 0.000116179388888889
TotalMinutes      : 0.00697076333333333
TotalSeconds      : 0.4182458
TotalMilliseconds : 418.2458
```

**Comment by amcasey** [comment]
AFAIK, I'm running Roslyn-Master-Signed-Release 20151112.1.

**Comment by jaredpar** [comment]
Andrew and I debugged further and we can take MSBuild out of the equation here by doing the following:

- Create a C# file that contains a hello world sample
- run `csc test.cs`: takes < 1 second
- run `csc /shared test.cs`: takes 30 seconds

This repros on the install dogfood builds but not the latest source code.

**Comment by jaredpar** [comment]
Have a good / bad pair of SHA's to examine.  Going through a bisect operation now.

**Comment by jaredpar** [comment]
This is fixed now in master.  Here are the set of steps that lead to this regression and fix:

The bug was introduced in commit bb15191681c28e387b9cb11cb810cb942f2e54de.  In particular from this block of code:

``` 
pipeStream.WaitForConnection();
CompilerServerLogger.Log("Pipe connection detected.");

if (!ClientAndOurIdentitiesMatch(pipeStream))
{
    var exception = new Exception("Client identity does not match server identity.");
    listenSource.SetException(exception);
    return;
}
```

The code in the `if` test fails 100% of the time if you haven't read data yet.  This check was moved here incorrectly.  As a result the server ignored every connection sent by a client.  

The key is it ignored but did not close.  In this case an exception is returned but the `pipeStream` value isn't closed.  This means the client gets no notification and ends up timing out after 30 seconds. 

This was fixed in commit 6a1fe0d9c6d26461e622e4e1b6eb3db427c3a67f in two ways:

1. The fix was moved to an appropriate place 
2. The test suite was updated such that it would have failed for the original commit
